src/storage/src/decode/mod.rs

```diff
@@ -492,7 +493,13 @@ pub fn render_decode_delimited<G: Scope, FromTime: Timestamp>(
                                 Some((decoder, buf)) => {
                                     decode_delimited(decoder, buf).await?.transpose()
                                 }
-                                None => None,
+                                None => {
+                                    if key_buf.is_some() {
```

Comment:
Looks like macthing on the key decoder needs to go inside the `Datum::Bytes(buf)` match arm above in line 486 instead of re-checking here